toolkit/tools/precacher/precacher.go

```diff
@@ -169,8 +168,11 @@ func monitorProgress(total int, results chan downloadResult, doneChannel chan st
 
 	for done := false; !done; {
 		// Wait for a result from a worker, or the done channel to be closed (which means all workers are done)
-		select {
-		case result := <-results:
+		result, ok := <-results
+		if !ok {
+			// All workers are done, finish this iteration of the loop and then return
+			done = true
+		} else {
```

Comment:
I think we still left the code a bit messy and are not utilizing what the language gives us.

Since we don't have `doneChannel` now, then we can also dump `done` and replace `for done := false; !done;` with `for result := range results`. It will exit on its own after reading the last result.